src/function/maybe_changed_after.rs

```diff
@@ -148,79 +201,95 @@ where
         );
 
         if self.shallow_verify_memo(db, zalsa, database_key_index, old_memo) {
-            return true;
+            return VerifyResult::Unchanged(Default::default());
         }
 
-        match &old_memo.revisions.origin {
-            QueryOrigin::Assigned(_) => {
-                // If the value was assigneed by another query,
-                // and that query were up-to-date,
-                // then we would have updated the `verified_at` field already.
-                // So the fact that we are here means that it was not specified
-                // during this revision or is otherwise stale.
-                //
-                // Example of how this can happen:
-                //
-                // Conditionally specified queries
-                // where the value is specified
-                // in rev 1 but not in rev 2.
-                return false;
-            }
-            QueryOrigin::BaseInput => {
-                // This value was `set` by the mutator thread -- ie, it's a base input and it cannot be out of date.
-                return true;
-            }
-            QueryOrigin::DerivedUntracked(_) => {
-                // Untracked inputs? Have to assume that it changed.
-                return false;
-            }
-            QueryOrigin::Derived(edges) => {
-                // Fully tracked inputs? Iterate over the inputs and check them, one by one.
-                //
-                // NB: It's important here that we are iterating the inputs in the order that
-                // they executed. It's possible that if the value of some input I0 is no longer
-                // valid, then some later input I1 might never have executed at all, so verifying
-                // it is still up to date is meaningless.
-                let last_verified_at = old_memo.verified_at.load();
-                for &(edge_kind, dependency_index) in edges.input_outputs.iter() {
-                    match edge_kind {
-                        EdgeKind::Input => {
-                            if dependency_index
-                                .maybe_changed_after(db.as_dyn_database(), last_verified_at)
-                            {
-                                return false;
+        loop {
+            let mut cycle_heads = FxHashSet::default();
```

Comment:
Is it intentional that we create a new `cycle_heads` in every iteration? Could we re-use the `cycle_heads` and instead call `clear` to avoid re-allocating the hash set on every iteration?